debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -212,20 +213,44 @@ record = bufferedEvent.getAndSet(record);
      */
     public void flushBuffer(Function<T, T> recordModifier) throws InterruptedException {
         assert buffering : "Unsupported for queues with disabled buffering";
+
         T record = bufferedEvent.getAndSet(null);
         if (record != null) {
+            // doEnqueue will check running state and throw InterruptedException if not running
             doEnqueue(recordModifier.apply(record));
         }
     }
 
     /**
-     * Disable buffering for the queue
+     * Disable buffering for the queue.
+     * This method enforces that the buffer must be empty (flushed) before disabling.
+     * Use this method in normal operation when you can guarantee the buffer is properly flushed.
+     * For cleanup scenarios where the buffer state might be inconsistent, use {@link #disableBufferingSafely()}.
+     *
+     * @throws AssertionError if the buffer is not empty (assertions enabled)
      */
     public void disableBuffering() {
         assert bufferedEvent.get() == null : "Buffer must be flushed";
         buffering = false;
     }
 
+    /**
+     * Safely disable buffering for the queue without throwing assertions.
+     * This method should be used in cleanup scenarios where the buffer state
+     * might be inconsistent due to shutdown timing or interruptions.
+     *
+     * @return true if buffering was disabled successfully, false if buffer was not empty
+     */
+    public boolean disableBufferingSafely() {
+        if (bufferedEvent.get() != null) {
+            LOGGER.warn("Disabling buffering with non-empty buffer - this may indicate incomplete flush during shutdown");
+            // Clear the buffer to prevent memory leaks
+            bufferedEvent.set(null);
+        }
+        buffering = false;
```

Comment:
I think this should go to the top of the method. Otherwise you can celan the buffer but another thread can fill it again before buffering is disabled.